File: java/shared/java/org/mochi/libmochi/zkgroup/auth/AuthCredentialPresentation.java
//
// Copyright 2020-2022 Mochi Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.mochi.libmochi.zkgroup.auth;

import static org.mochi.libmochi.internal.FilterExceptions.filterExceptions;

import java.time.Instant;
import org.mochi.libmochi.internal.Native;
import org.mochi.libmochi.zkgroup.InvalidInputException;
import org.mochi.libmochi.zkgroup.groups.UuidCiphertext;
import org.mochi.libmochi.zkgroup.internal.ByteArray;

public final class AuthCredentialPresentation extends ByteArray {

  public enum Version {
    V1,
    V2,
    V3,
    V4,
    UNKNOWN
  };

  public AuthCredentialPresentation(byte[] contents) throws InvalidInputException {
    super(contents);
    filterExceptions(
        InvalidInputException.class,
        () -> Native.AuthCredentialPresentation_CheckValidContents(contents));
  }

  public UuidCiphertext getUuidCiphertext() {
    byte[] newContents = Native.AuthCredentialPresentation_GetUuidCiphertext(contents);

    try {
      return new UuidCiphertext(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  /** Returns the PNI ciphertext for this credential. May be {@code null}. */
  public UuidCiphertext getPniCiphertext() {
    byte[] newContents = Native.AuthCredentialPresentation_GetPniCiphertext(contents);
    if (newContents == null) {
      return null;
    }

    try {
      return new UuidCiphertext(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public Instant getRedemptionTime() {
    return Instant.ofEpochSecond(Native.AuthCredentialPresentation_GetRedemptionTime(contents));
  }

  public Version getVersion() {
    byte version = this.contents[0];
    final Version[] values = Version.values();
    if (version < values.length) {
      return values[version];
    }
    return Version.UNKNOWN;
  }
}

File: swift/Sources/LibMochiClient/zkgroup/AuthCredentialPresentation.swift
//
// Copyright 2020-2022 Mochi Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import MochiFfi

public class AuthCredentialPresentation: ByteArray {
    public required init(contents: [UInt8]) throws {
        try super.init(contents, checkValid: mochi_auth_credential_presentation_check_valid_contents)
    }

    public func getUuidCiphertext() throws -> UuidCiphertext {
        return try withUnsafeBorrowedBuffer { buffer in
            try invokeFnReturningSerialized {
                mochi_auth_credential_presentation_get_uuid_ciphertext($0, buffer)
            }
        }
    }

    public func getPniCiphertext() throws -> UuidCiphertext? {
        return try withUnsafeBorrowedBuffer { buffer in
            try invokeFnReturningOptionalVariableLengthSerialized {
                mochi_auth_credential_presentation_get_pni_ciphertext_or_empty($0, buffer)
            }
        }
    }

    public func getRedemptionTime() throws -> Date {
        let secondsSinceEpoch = try withUnsafeBorrowedBuffer { buffer in
            try invokeFnReturningInteger {
                mochi_auth_credential_presentation_get_redemption_time($0, buffer)
            }
        }
        return Date(timeIntervalSince1970: TimeInterval(secondsSinceEpoch))
    }
}
